[MathML] Render empty <msqrt> radicals correctly instead of skipping paint due to firstChild()

Ahmad-S792 opened the pull request:
#### db88908da038d49d7e85d37692fced0d47363227
<pre>
[MathML] Render empty &lt;msqrt&gt; radicals correctly instead of skipping paint due to firstChild()
<a href="https://bugs.webkit.org/show_bug.cgi?id=302563">https://bugs.webkit.org/show_bug.cgi?id=302563</a>
<a href="https://rdar.apple.com/164776629">rdar://164776629</a>

Reviewed by Frédéric Wang.

Previously, RenderMathMLRoot::paint() returned early if firstChild() was null.
This prevented empty &lt;msqrt&gt; elements from painting, which is invalid per
MathML Core. Now we remove the firstChild() check and rely solely on isValid()
to guard painting. Empty &lt;msqrt&gt; elements now correctly render the radical
glyph with zero-length base. This preserves remaining checks and early returns
and keep them unchanged. As per MathML Core, &lt;mroot&gt; is valid if and only if
it has exactly two in-flow children (which isValid()) handles while &lt;msqrt&gt;
are always valid.

This is tested by Web Platform Test (empty-msqrt.html) but unfortuantely
due to still some rendering differences, we cannot unskip it. Although,
rather than being empty, it now renders. It fails with 0.02% difference.

WPT Test: mathml/presentation-markup/radicals/empty-msqrt.html

* Source/WebCore/rendering/mathml/RenderMathMLRoot.cpp:
(WebCore::RenderMathMLRoot::paint):

Canonical link: <a href="https://commits.webkit.org/303134@main">https://commits.webkit.org/303134@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/59522f3a4025cce4a9e72a418d9be7e8d6abd2c8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/131457 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/3789 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/42422 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/138963 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/83231 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/2b85b725-cc8a-4cb2-a1e8-82a3a8bcb2e1) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/133327 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/3811 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/3629 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/100381 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/83231 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/bf2448a8-2db6-48ba-b1f8-79acd2889cdf) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/134403 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/2750 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/117667 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/81168 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/a98c8520-5911-40b3-8e28-e6a1963ac450) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/2665 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/411 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/82155 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/111286 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/35799 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/141608 "Built successfully") | | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/153/builds/3532 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/36308 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/108746 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/160/builds/3580 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/3156 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/108967 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27603 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-26-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/113995 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/56721 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/158/builds/3594 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/32411 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/3419 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/67005 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/3620 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/152/builds/3524 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

fred-wang reviewed (approved):
Thanks. That looks an obsolete check indeed.

I wonder if in your commit message you can explicitly mention `<mroot>` which is "valid" iff it has exactly two in-flow children. `<msqrt>` are always "valid" per MathML Core now.

I'm also curious why `msqrt-empty` is actually failing? (Probably we should have a follow-up bug for that)

Ahmad-S792 committed:
[MathML] Render empty <msqrt> radicals correctly instead of skipping paint due to firstChild()
https://bugs.webkit.org/show_bug.cgi?id=302563
rdar://164776629

Reviewed by Frédéric Wang.

Previously, RenderMathMLRoot::paint() returned early if firstChild() was null.
This prevented empty <msqrt> elements from painting, which is invalid per
MathML Core. Now we remove the firstChild() check and rely solely on isValid()
to guard painting. Empty <msqrt> elements now correctly render the radical
glyph with zero-length base. This preserves remaining checks and early returns
and keep them unchanged. As per MathML Core, <mroot> is valid if and only if
it has exactly two in-flow children (which isValid()) handles while <msqrt>
are always valid.

This is tested by Web Platform Test (empty-msqrt.html) but unfortuantely
due to still some rendering differences, we cannot unskip it. Although,
rather than being empty, it now renders. It fails with 0.02% difference.

WPT Test: mathml/presentation-markup/radicals/empty-msqrt.html

* Source/WebCore/rendering/mathml/RenderMathMLRoot.cpp:
(WebCore::RenderMathMLRoot::paint):

Canonical link: https://commits.webkit.org/303134@main
